### DISCUSSION TITLE
Kernel TotalLagrangianStressDivergence and DerivativeParsedMaterial

### DISCUSSION THREAD

**Discussion by rcontinuum** [idaholab/moose#21749] [Q&A Modules: Solid mechanics] [answered]
Hello, 
I'm struggling a bit with the new `TotalLagrangianStressDivergence` kernel. I want to couple this kernel to a Phase Field Kernel which uses/accesses a `DerivativeParsedMaterial` (e.g., like in `crack2d_iso.i` from Phase Field Fracture) 

```
[Materials]
  [degradation]
    type = DerivativeParsedMaterial
    f_name = 'degradation'
    args = 'c'
    function = '(1.0-c)^2'
    derivative_order = 2
  []
  ...
```
To retrieve the first derivative of this Material function I need to do something like (see `ComputePFFractureStressBase.C`, old TensorMechanics)
```
ComputePFFractureStressBase::ComputePFFractureStressBase(const InputParameters & parameters)
  : ComputeStressBase(parameters),
  ...
  _dDdc(getMaterialPropertyDerivative<Real>("D_name", getVar("c", 0)->name())),   // get first derivative wrt 'c'
```

Now, in   `ComputeStressBase.h` (old TensorMechanics) we have

`class ComputeStressBase : public DerivativeMaterialInterface<Material>`

whereas in `ComputeLagrangianStressBase.h` (new TensorMechanics)  we have 

`class ComputeLagrangianStressBase : public Material`

If I understand correct, but I have little experience with MOOSE, then this `: public Material` prevents me from accessing the first derivative by  `getMaterialPropertyDerivative<Real>`

I could try to change in `ComputeLagrangianStressBase.h` to
`class ComputeLagrangianStressBase : public DerivativeMaterialInterface<Material>`

and in `ComputeLagrangianStressBase.C` change `***Material***` 
```
ComputeLagrangianStressBase::ComputeLagrangianStressBase(const InputParameters & parameters)
: ***Material***(parameters),
```
but before I go this route (and "get lost") I like to ask the experts in the field for some advise.

Cheers and thank you

**Comment by hugary1995** [comment, marked as answer]
I have plan to change add derivative interface in the new stress calculators, so this will work out-of-the-box in the future.

Meanwhile, you can add the derivative interface decorator anywhere you like, i.e. in the class you are actually retrieving the derivatives.

**Reply by rcontinuum** [reply]
> I have plan to change add derivative interface in the new stress calculators, so this will work out-of-the-box in the future.

That sounds good. It's not really urgent, so I simply wait.

Meanwhile I just "hard code" the functions and their derivatives into the material model. But I thought `DerivativeParsedMaterial` would give me a bit more flexibility in the input file. 

Thank you

**Reply by hugary1995** [reply]
Yeah, my point was you don't have to wait. I assume you have something like this in your stress calculator:
```cpp
class MyStressCalculator : public ParentStressCalculator
```
simply change it to
```cpp
class MyStressCalculator : public DerivativeMaterialInterface<ParentStressCalculator>
```
and you'll get all the interface apis.

**Reply by rcontinuum** [reply]
Ah, this is how it works. 
Thanks

**Comment by GiudGiud** [comment]
Any reason you can't get the material property derivative as a material property directly, with the "dCdXXX" name directly?
Sorry if this doesnt make sense, not necessarily super familiar with this system
